packages/editor/src/lib/editor/managers/FontManager.ts

```diff
@@ -0,0 +1,201 @@
+			`  src: ${font.src};`,
+			fontDisplay ? `  font-display: ${fontDisplay};` : null,
+			font.ascentOverride ? `  ascent-override: ${font.ascentOverride};` : null,
+			font.descentOverride ? `  descent-override: ${font.descentOverride};` : null,
+			font.stretch ? `  font-stretch: ${font.stretch};` : null,
+			font.style ? `  font-style: ${font.style};` : null,
+			font.weight ? `  font-weight: ${font.weight};` : null,
+			font.featureSettings ? `  font-feature-settings: ${font.featureSettings};` : null,
+			font.lineGapOverride ? `  line-gap-override: ${font.lineGapOverride};` : null,
+			`}`,
+		]).join('\n')
+	}
+
+	private fontStates = new Map<TLFontFace, Atom<FontState>>()
+
+	constructor(
+		private readonly editor: Editor,
+		private readonly assetUrls?: Record<string, string>
+	) {}
+
+	@computed private getShapeFontFacesCache() {
+		return createComputedCache('shape font faces', (editor: Editor, shape: TLShape) => {
+			const shapeUtil = editor.getShapeUtil(shape)
+			return shapeUtil.getFontFaces(shape)
+		})
+	}
+
+	getShapeFontFaces(shape: TLShape): TLFontFace[] {
+		return this.getShapeFontFacesCache().get(this.editor, shape.id) ?? EMPTY_ARRAY
+	}
+
+	@computed private shapeFontLoadStateCache() {
+		return createComputedCache('shape font load state', (editor: Editor, shape: TLShape) => {
+			const states = this.getShapeFontFaces(shape).map((face) => this.getFontState(face))
+			return states
+		})
+	}
+	trackFontsForShape(shape: TLShape) {
+		this.shapeFontLoadStateCache().get(this.editor, shape.id)
+	}
+
+	async loadRequiredFontsForCurrentPage(limit = Infinity) {
+		const neededFonts = new Set<TLFontFace>()
+		for (const shapeId of this.editor.getCurrentPageShapeIds()) {
+			for (const font of this.getShapeFontFaces(this.editor.getShape(shapeId)!)) {
+				neededFonts.add(font)
+			}
+		}
+
+		if (neededFonts.size > limit) {
+			return
+		}
+
+		const promises = Array.from(neededFonts, (font) => this.ensureFontIsLoaded(font))
+		await Promise.all(promises)
+	}
+
+	private missingFontAtom = atom('missing font tracker', 0)
+	private getFontState(font: TLFontFace): FontState | null {
+		const stateAtom = this.fontStates.get(font)
+		if (!stateAtom) {
+			// if we don't have a state for this font, listen to the missing font atom. When we add
+			// a font, this will change and cause this to re-evaluate.
+			this.missingFontAtom.get()
+			return null
+		}
+		return stateAtom.get()
+	}
+
+	ensureFontIsLoaded(font: TLFontFace): Promise<void> {
+		const state = this.getFontState(font)
+		if (state) return state.loadingPromise
+
+		const url = this.assetUrls?.[font.src.url] ?? font.src.url
+		const instance = new FontFace(font.fontFamily, `url(${JSON.stringify(url)})`, {
+			ascentOverride: font.ascentOverride,
+			descentOverride: font.descentOverride,
+			stretch: font.stretch,
+			style: font.style,
+			weight: font.weight,
+			featureSettings: font.featureSettings,
+			lineGapOverride: font.lineGapOverride,
+			display: 'swap',
```

Comment:
let's consider `block` here instead of `swap`

Reply:
I was unable to see a noticeable difference for these. I think they're maybe intended for a different use-case than ours